File: libs/php/src/ConvertorioClient.php
<?php

namespace Convertorio\SDK;

use Exception;

/**
 * Convertorio SDK Client for PHP
 *
 * Simple and powerful SDK to convert images using the Convertorio API.
 *
 * @package Convertorio\SDK
 * @version 1.2.0
 */
class ConvertorioClient
{
    /**
     * @var string API key
     */
    private $apiKey;

    /**
     * @var string Base URL for API
     */
    private $baseUrl;

    /**
     * @var array Event callbacks
     */
    private $eventCallbacks = [];

    /**
     * @var bool Verify SSL certificates
     */
    private $verifySsl = true;

    /**
     * Create a new Convertorio client
     *
     * @param string $apiKey Your Convertorio API key
     * @param string|null $baseUrl Optional custom API base URL
     * @param bool $verifySsl Verify SSL certificates (set to false for development)
     * @throws Exception If API key is not provided
     */
    public function __construct(string $apiKey, ?string $baseUrl = null, bool $verifySsl = true)
    {
        if (empty($apiKey)) {
            throw new Exception('API key is required. Get yours at https://convertorio.com/account');
        }

        $this->apiKey = $apiKey;
        $this->baseUrl = $baseUrl ?? 'https://api.convertorio.com';
        $this->verifySsl = $verifySsl;
    }

    /**
     * Register an event callback
     *
     * @param string $event Event name (start, progress, status, complete, error)
     * @param callable $callback Callback function
     * @return self
     */
    public function on(string $event, callable $callback): self
    {
        if (!isset($this->eventCallbacks[$event])) {
            $this->eventCallbacks[$event] = [];
        }
        $this->eventCallbacks[$event][] = $callback;
        return $this;
    }

    /**
     * Emit an event to registered callbacks
     *
     * @param string $event Event name
     * @param array $data Event data
     */
    private function emit(string $event, array $data): void
    {
        if (isset($this->eventCallbacks[$event])) {
            foreach ($this->eventCallbacks[$event] as $callback) {
                call_user_func($callback, $data);
            }
        }
    }

    /**
     * Convert an image file
     *
     * @param string $inputPath Path to the input file
     * @param string $targetFormat Target format (jpg, png, webp, avif, etc.)
     * @param string|null $outputPath Optional output path (auto-generated if not provided)
     * @param array|null $conversionMetadata Advanced conversion options
     *   - aspect_ratio: Target aspect ratio (original, 1:1, 4:3, 16:9, 9:16, 21:9, custom)
     *   - crop_strategy: Crop strategy (fit, crop-center, crop-top, crop-bottom, crop-left, crop-right)
     *   - quality: Compression quality 1-100 (for JPG, WebP, AVIF, HEIC)
     *   - icon_size: Icon size in pixels (for ICO format: 16, 32, 48, 64, 128, 256)
     *   - resize_width: Target width in pixels (1-10000)
     *   - resize_height: Target height in pixels (1-10000)
     * @return array Conversion result with download URL and file path
     * @throws Exception If conversion fails
     */
    public function convertFile(
        string $inputPath,
        string $targetFormat,
        ?string $outputPath = null,
        ?array $conversionMetadata = null
    ): array {
        // Validate input file exists
        if (!file_exists($inputPath)) {
            throw new Exception("Input file not found: {$inputPath}");
        }

        $fileSize = filesize($inputPath);
        $fileName = basename($inputPath);
        $sourceFormat = strtolower(pathinfo($inputPath, PATHINFO_EXTENSION));

        $this->emit('start', [
            'file_name' => $fileName,
            'source_format' => $sourceFormat,
            'target_format' => $targetFormat
        ]);

        try {
            // Step 1: Request upload URL
            $this->emit('progress', [
                'step' => 'requesting-upload-url',
                'message' => 'Requesting upload URL from server...'
            ]);

            $requestBody = [
                'filename' => $fileName,
                'source_format' => $sourceFormat,
                'target_format' => strtolower($targetFormat),
                'file_size' => $fileSize
            ];

            // Add conversion metadata if provided
            if ($conversionMetadata !== null && !empty($conversionMetadata)) {
                $requestBody['conversion_metadata'] = $conversionMetadata;
            }

            $uploadRequest = $this->makeRequest('POST', '/v1/convert/upload-url', $requestBody);

            if (!($uploadRequest['success'] ?? false)) {
                throw new Exception($uploadRequest['error'] ?? 'Failed to get upload URL');
            }

            $jobId = $uploadRequest['job_id'];
            $uploadUrl = $uploadRequest['upload_url'];

            // Step 2: Upload file to S3
            $this->emit('progress', [
                'step' => 'uploading',
                'message' => 'Uploading file to cloud storage...',
                'job_id' => $jobId
            ]);

            $fileData = file_get_contents($inputPath);
            $this->uploadFile($uploadUrl, $fileData, "image/{$sourceFormat}");

            // Step 3: Confirm upload and queue conversion
            $this->emit('progress', [
                'step' => 'confirming',
                'message' => 'Confirming upload and queuing conversion...',
                'job_id' => $jobId
            ]);

            $confirmRequest = $this->makeRequest('POST', '/v1/convert/confirm', ['job_id' => $jobId]);

            if (!($confirmRequest['success'] ?? false)) {
                throw new Exception($confirmRequest['error'] ?? 'Failed to confirm upload');
            }

            // Step 4: Poll for completion
            $this->emit('progress', [
                'step' => 'converting',
                'message' => 'Converting image...',
                'job_id' => $jobId,
                'status' => $confirmRequest['status'] ?? 'queued'
            ]);

            $result = $this->pollJobStatus($jobId);

            // Step 5: Download converted file
            $this->emit('progress', [
                'step' => 'downloading',
                'message' => 'Downloading converted file...',
                'job_id' => $jobId
            ]);

            $finalOutputPath = $outputPath ?? $this->generateOutputPath($inputPath, $targetFormat);
            $this->downloadFile($result['download_url'], $finalOutputPath);

            // Build conversion result
            $conversionResult = [
                'success' => true,
                'job_id' => $jobId,
                'input_path' => $inputPath,
                'output_path' => $finalOutputPath,
                'source_format' => $sourceFormat,
                'target_format' => strtolower($targetFormat),
                'file_size' => filesize($finalOutputPath),
                'processing_time' => $result['processing_time_ms'] ?? 0,
                'download_url' => $result['download_url']
            ];

            $this->emit('complete', $conversionResult);
            return $conversionResult;

        } catch (Exception $error) {
            $errorDetails = [
                'success' => false,
                'error' => $error->getMessage(),
                'input_path' => $inputPath,
                'target_format' => $targetFormat
            ];
            $this->emit('error', $errorDetails);
            throw $error;
        }
    }

    /**
     * Poll job status until completion
     *
     * @param string $jobId Job ID to poll
     * @param int $maxAttempts Maximum polling attempts
     * @param int $interval Polling interval in seconds
     * @return array Job result
     * @throws Exception If job fails or times out
     */
    private function pollJobStatus(string $jobId, int $maxAttempts = 60, int $interval = 2): array
    {
        $attempts = 0;

        while ($attempts < $maxAttempts) {
            $attempts++;

            // Wait before polling (except first attempt)
            if ($attempts > 1) {
                sleep($interval);
            }

            $statusRequest = $this->makeRequest('GET', "/v1/jobs/{$jobId}");

            if (!($statusRequest['success'] ?? false)) {
                throw new Exception('Failed to get job status');
            }

            $job = $statusRequest['job'];
            $status = $job['status'];

            $this->emit('status', [
                'job_id' => $jobId,
                'status' => $status,
                'attempt' => $attempts,
                'max_attempts' => $maxAttempts
            ]);

            if ($status === 'completed') {
                return $job;
            }

            if ($status === 'failed') {
                throw new Exception($job['error_message'] ?? 'Conversion failed');
            }

            if ($status === 'expired') {
                throw new Exception('Job expired');
            }
        }

        throw new Exception('Conversion timeout - job did not complete in time');
    }

    /**
     * Download file from URL
     *
     * @param string $url Download URL
     * @param string $outputPath Output file path
     * @throws Exception If download fails
     */
    private function downloadFile(string $url, string $outputPath): void
    {
        // Use cURL for better SSL control
        $ch = curl_init($url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);

        // SSL verification settings
        if (!$this->verifySsl) {
            curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
            curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);
        }

        $fileData = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $error = curl_error($ch);
        curl_close($ch);

        if ($error) {
            throw new Exception("Download failed: {$error}");
        }

        if ($httpCode < 200 || $httpCode >= 300) {
            throw new Exception("Failed to download file. HTTP Status: {$httpCode}");
        }

        if ($fileData === false) {
            throw new Exception("Failed to download file from: {$url}");
        }

        // Ensure output directory exists
        $outputDir = dirname($outputPath);
        if (!is_dir($outputDir)) {
            mkdir($outputDir, 0755, true);
        }

        if (file_put_contents($outputPath, $fileData) === false) {
            throw new Exception("Failed to write file to: {$outputPath}");
        }
    }

    /**
     * Upload file to S3
     *
     * @param string $url Upload URL
     * @param string $fileData File data
     * @param string $contentType Content type
     * @throws Exception If upload fails
     */
    private function uploadFile(string $url, string $fileData, string $contentType): void
    {
        $ch = curl_init($url);
        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'PUT');
        curl_setopt($ch, CURLOPT_POSTFIELDS, $fileData);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_HTTPHEADER, [
            "Content-Type: {$contentType}",
            'Content-Length: ' . strlen($fileData)
        ]);

        // SSL verification settings
        if (!$this->verifySsl) {
            curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
            curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);
        }

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $error = curl_error($ch);
        curl_close($ch);

        if ($error) {
            throw new Exception("Upload failed: {$error}");
        }

        if ($httpCode < 200 || $httpCode >= 300) {
            throw new Exception("Failed to upload file. HTTP Status: {$httpCode}");
        }
    }

    /**
     * Generate output path based on input path and target format
     *
     * @param string $inputPath Input file path
     * @param string $targetFormat Target format
     * @return string Output file path
     */
    private function generateOutputPath(string $inputPath, string $targetFormat): string
    {
        $dir = dirname($inputPath);
        $baseName = pathinfo($inputPath, PATHINFO_FILENAME);
        return $dir . DIRECTORY_SEPARATOR . $baseName . '.' . strtolower($targetFormat);
    }

    /**
     * Make HTTP request to API
     *
     * @param string $method HTTP method
     * @param string $endpoint API endpoint
     * @param array|null $data Request data
     * @return array Response data
     * @throws Exception If request fails
     */
    private function makeRequest(string $method, string $endpoint, ?array $data = null): array
    {
        $url = $this->baseUrl . $endpoint;
        $ch = curl_init($url);

        $headers = [
            'Authorization: Bearer ' . $this->apiKey,
            'Content-Type: application/json'
        ];

        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);

        // SSL verification settings
        if (!$this->verifySsl) {
            curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
            curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);
        }

        if ($method === 'POST') {
            curl_setopt($ch, CURLOPT_POST, true);
            if ($data !== null) {
                curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));
            }
        } elseif ($method === 'GET' && $data !== null) {
            $url .= '?' . http_build_query($data);
            curl_setopt($ch, CURLOPT_URL, $url);
        }

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $error = curl_error($ch);
        curl_close($ch);

        if ($error) {
            throw new Exception("Request failed: {$error}");
        }

        if ($httpCode < 200 || $httpCode >= 300) {
            throw new Exception("HTTP request failed with status code: {$httpCode}");
        }

        $decoded = json_decode($response, true);
        if ($decoded === null) {
            throw new Exception("Failed to decode JSON response");
        }

        return $decoded;
    }

    /**
     * Get account information
     *
     * @return array Account details
     * @throws Exception If request fails
     */
    public function getAccount(): array
    {
        $response = $this->makeRequest('GET', '/v1/account');

        if (!($response['success'] ?? false)) {
            throw new Exception($response['error'] ?? 'Failed to get account info');
        }

        return $response['account'];
    }

    /**
     * List conversion jobs
     *
     * @param int $limit Number of jobs to return (max: 100)
     * @param int $offset Offset for pagination
     * @param string|null $status Filter by status (completed, failed, processing, etc.)
     * @return array List of jobs
     * @throws Exception If request fails
     */
    public function listJobs(int $limit = 50, int $offset = 0, ?string $status = null): array
    {
        $params = ['limit' => $limit, 'offset' => $offset];
        if ($status !== null) {
            $params['status'] = $status;
        }

        $response = $this->makeRequest('GET', '/v1/jobs', $params);

        if (!($response['success'] ?? false)) {
            throw new Exception($response['error'] ?? 'Failed to list jobs');
        }

        return $response['jobs'];
    }

    /**
     * Get job status
     *
     * @param string $jobId Job ID
     * @return array Job details
     * @throws Exception If request fails
     */
    public function getJob(string $jobId): array
    {
        $response = $this->makeRequest('GET', "/v1/jobs/{$jobId}");

        if (!($response['success'] ?? false)) {
            throw new Exception($response['error'] ?? 'Failed to get job');
        }

        return $response['job'];
    }
}

File: libs/java/src/main/java/com/sedesoft/convertorio/ConvertorioClient.java
package com.sedesoft.convertorio;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import okhttp3.*;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

/**
 * Convertorio SDK Client for Java
 *
 * Simple and powerful SDK to convert images using the Convertorio API.
 *
 * @version 1.2.0
 */
public class ConvertorioClient {
    private final String apiKey;
    private final String baseUrl;
    private final OkHttpClient httpClient;
    private final Gson gson;
    private final Map<String, Consumer<EventData>> eventListeners;

    private static final int MAX_POLLING_ATTEMPTS = 60;
    private static final int POLLING_INTERVAL_MS = 2000;
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");

    /**
     * Create a new Convertorio client
     *
     * @param config Configuration object
     * @throws IllegalArgumentException if API key is not provided
     */
    public ConvertorioClient(ClientConfig config) {
        if (config.getApiKey() == null || config.getApiKey().isEmpty()) {
            throw new IllegalArgumentException("API key is required. Get yours at https://convertorio.com/account");
        }

        this.apiKey = config.getApiKey();
        this.baseUrl = config.getBaseUrl() != null ? config.getBaseUrl() : "https://api.convertorio.com";
        this.gson = new Gson();
        this.eventListeners = new HashMap<>();

        this.httpClient = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build();
    }

    /**
     * Register an event listener
     *
     * @param event Event name (start, progress, status, complete, error)
     * @param listener Event callback
     * @return this client instance for chaining
     */
    public ConvertorioClient on(String event, Consumer<EventData> listener) {
        eventListeners.put(event, listener);
        return this;
    }

    /**
     * Emit an event to registered listeners
     */
    private void emit(String event, EventData data) {
        Consumer<EventData> listener = eventListeners.get(event);
        if (listener != null) {
            listener.accept(data);
        }
    }

    /**
     * Convert an image file
     *
     * @param options Conversion options
     * @return Conversion result with download URL and file path
     * @throws ConvertorioException if conversion fails
     */
    public ConversionResult convertFile(ConversionOptions options) throws ConvertorioException {
        if (options.getInputPath() == null) {
            throw new IllegalArgumentException("inputPath is required");
        }

        if (options.getTargetFormat() == null) {
            throw new IllegalArgumentException("targetFormat is required");
        }

        File inputFile = new File(options.getInputPath());
        if (!inputFile.exists()) {
            throw new ConvertorioException("Input file not found: " + options.getInputPath());
        }

        String fileName = inputFile.getName();
        String sourceFormat = getFileExtension(fileName);
        long fileSize = inputFile.length();

        emit("start", EventData.builder()
            .set("fileName", fileName)
            .set("sourceFormat", sourceFormat)
            .set("targetFormat", options.getTargetFormat())
            .build());

        try {
            // Step 1: Request upload URL
            emit("progress", EventData.builder()
                .set("step", "requesting-upload-url")
                .set("message", "Requesting upload URL from server...")
                .build());

            JsonObject requestBody = new JsonObject();
            requestBody.addProperty("filename", fileName);
            requestBody.addProperty("source_format", sourceFormat);
            requestBody.addProperty("target_format", options.getTargetFormat().toLowerCase());
            requestBody.addProperty("file_size", fileSize);

            if (options.getConversionMetadata() != null && !options.getConversionMetadata().isEmpty()) {
                requestBody.add("conversion_metadata", gson.toJsonTree(options.getConversionMetadata()));
            }

            JsonObject uploadResponse = makeRequest("POST", "/v1/convert/upload-url", requestBody);

            if (!uploadResponse.get("success").getAsBoolean()) {
                throw new ConvertorioException(uploadResponse.has("error") ?
                    uploadResponse.get("error").getAsString() : "Failed to get upload URL");
            }

            String jobId = uploadResponse.get("job_id").getAsString();
            String uploadUrl = uploadResponse.get("upload_url").getAsString();

            // Step 2: Upload file to S3
            emit("progress", EventData.builder()
                .set("step", "uploading")
                .set("message", "Uploading file to cloud storage...")
                .set("jobId", jobId)
                .build());

            uploadFile(uploadUrl, inputFile, sourceFormat);

            // Step 3: Confirm upload and queue conversion
            emit("progress", EventData.builder()
                .set("step", "confirming")
                .set("message", "Confirming upload and queuing conversion...")
                .set("jobId", jobId)
                .build());

            JsonObject confirmBody = new JsonObject();
            confirmBody.addProperty("job_id", jobId);
            JsonObject confirmResponse = makeRequest("POST", "/v1/convert/confirm", confirmBody);

            if (!confirmResponse.get("success").getAsBoolean()) {
                throw new ConvertorioException(confirmResponse.has("error") ?
                    confirmResponse.get("error").getAsString() : "Failed to confirm upload");
            }

            // Step 4: Poll for completion
            emit("progress", EventData.builder()
                .set("step", "converting")
                .set("message", "Converting image...")
                .set("jobId", jobId)
                .set("status", confirmResponse.has("status") ? confirmResponse.get("status").getAsString() : "queued")
                .build());

            Job job = pollJobStatus(jobId);

            // Step 5: Download converted file
            emit("progress", EventData.builder()
                .set("step", "downloading")
                .set("message", "Downloading converted file...")
                .set("jobId", jobId)
                .build());

            String outputPath = options.getOutputPath() != null ?
                options.getOutputPath() : generateOutputPath(options.getInputPath(), options.getTargetFormat());

            downloadFile(job.getDownloadUrl(), outputPath);

            ConversionResult result = new ConversionResult(
                true,
                jobId,
                options.getInputPath(),
                outputPath,
                sourceFormat,
                options.getTargetFormat().toLowerCase(),
                new File(outputPath).length(),
                job.getProcessingTimeMs(),
                job.getDownloadUrl(),
                job.getTokensUsed()
            );

            emit("complete", EventData.fromResult(result));
            return result;

        } catch (ConvertorioException e) {
            EventData errorData = EventData.builder()
                .set("success", false)
                .set("error", e.getMessage())
                .set("inputPath", options.getInputPath())
                .set("targetFormat", options.getTargetFormat())
                .build();
            emit("error", errorData);
            throw e;
        }
    }

    /**
     * Poll job status until completion
     */
    private Job pollJobStatus(String jobId) throws ConvertorioException {
        int attempts = 0;

        while (attempts < MAX_POLLING_ATTEMPTS) {
            attempts++;

            if (attempts > 1) {
                try {
                    Thread.sleep(POLLING_INTERVAL_MS);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    throw new ConvertorioException("Polling interrupted");
                }
            }

            JsonObject statusResponse = makeRequest("GET", "/v1/jobs/" + jobId, null);

            if (!statusResponse.get("success").getAsBoolean()) {
                throw new ConvertorioException("Failed to get job status");
            }

            Job job = gson.fromJson(statusResponse.get("job"), Job.class);
            String status = job.getStatus();

            emit("status", EventData.builder()
                .set("jobId", jobId)
                .set("status", status)
                .set("attempt", attempts)
                .set("maxAttempts", MAX_POLLING_ATTEMPTS)
                .build());

            if ("completed".equals(status)) {
                return job;
            }

            if ("failed".equals(status)) {
                throw new ConvertorioException(job.getErrorMessage() != null ?
                    job.getErrorMessage() : "Conversion failed");
            }

            if ("expired".equals(status)) {
                throw new ConvertorioException("Job expired");
            }
        }

        throw new ConvertorioException("Conversion timeout - job did not complete in time");
    }

    /**
     * Upload file to S3
     */
    private void uploadFile(String url, File file, String sourceFormat) throws ConvertorioException {
        try {
            byte[] fileData = Files.readAllBytes(file.toPath());
            RequestBody body = RequestBody.create(fileData, MediaType.get("image/" + sourceFormat));

            Request request = new Request.Builder()
                .url(url)
                .put(body)
                .addHeader("Content-Type", "image/" + sourceFormat)
                .build();

            try (Response response = httpClient.newCall(request).execute()) {
                if (!response.isSuccessful()) {
                    throw new ConvertorioException("Failed to upload file. HTTP Status: " + response.code());
                }
            }
        } catch (IOException e) {
            throw new ConvertorioException("Upload failed: " + e.getMessage(), e);
        }
    }

    /**
     * Download file from URL
     */
    private void downloadFile(String url, String outputPath) throws ConvertorioException {
        try {
            Request request = new Request.Builder()
                .url(url)
                .get()
                .build();

            try (Response response = httpClient.newCall(request).execute()) {
                if (!response.isSuccessful()) {
                    throw new ConvertorioException("Failed to download file. HTTP Status: " + response.code());
                }

                File outputFile = new File(outputPath);
                File parentDir = outputFile.getParentFile();
                if (parentDir != null && !parentDir.exists()) {
                    parentDir.mkdirs();
                }

                try (InputStream inputStream = response.body().byteStream();
                     FileOutputStream outputStream = new FileOutputStream(outputFile)) {
                    byte[] buffer = new byte[8192];
                    int bytesRead;
                    while ((bytesRead = inputStream.read(buffer)) != -1) {
                        outputStream.write(buffer, 0, bytesRead);
                    }
                }
            }
        } catch (IOException e) {
            throw new ConvertorioException("Download failed: " + e.getMessage(), e);
        }
    }

    /**
     * Make HTTP request to API
     */
    private JsonObject makeRequest(String method, String endpoint, JsonObject data) throws ConvertorioException {
        try {
            String url = baseUrl + endpoint;
            Request.Builder requestBuilder = new Request.Builder()
                .url(url)
                .addHeader("Authorization", "Bearer " + apiKey)
                .addHeader("Content-Type", "application/json");

            if ("POST".equals(method) && data != null) {
                RequestBody body = RequestBody.create(gson.toJson(data), JSON);
                requestBuilder.post(body);
            } else {
                requestBuilder.get();
            }

            try (Response response = httpClient.newCall(requestBuilder.build()).execute()) {
                if (!response.isSuccessful()) {
                    throw new ConvertorioException("HTTP request failed with status code: " + response.code());
                }

                String responseBody = response.body().string();
                return gson.fromJson(responseBody, JsonObject.class);
            }
        } catch (IOException e) {
            throw new ConvertorioException("Request failed: " + e.getMessage(), e);
        }
    }

    /**
     * Generate output path based on input path and target format
     */
    private String generateOutputPath(String inputPath, String targetFormat) {
        File inputFile = new File(inputPath);
        String dir = inputFile.getParent();
        String baseName = getFileNameWithoutExtension(inputFile.getName());
        String separator = dir != null ? File.separator : "";
        return (dir != null ? dir : "") + separator + baseName + "." + targetFormat.toLowerCase();
    }

    /**
     * Get file extension without dot
     */
    private String getFileExtension(String fileName) {
        int lastDot = fileName.lastIndexOf('.');
        if (lastDot > 0) {
            return fileName.substring(lastDot + 1).toLowerCase();
        }
        return "";
    }

    /**
     * Get file name without extension
     */
    private String getFileNameWithoutExtension(String fileName) {
        int lastDot = fileName.lastIndexOf('.');
        if (lastDot > 0) {
            return fileName.substring(0, lastDot);
        }
        return fileName;
    }

    /**
     * Get account information
     *
     * @return Account details
     * @throws ConvertorioException if request fails
     */
    public Account getAccount() throws ConvertorioException {
        JsonObject response = makeRequest("GET", "/v1/account", null);

        if (!response.get("success").getAsBoolean()) {
            throw new ConvertorioException(response.has("error") ?
                response.get("error").getAsString() : "Failed to get account info");
        }

        return gson.fromJson(response.get("account"), Account.class);
    }

    /**
     * List conversion jobs
     *
     * @param limit Number of jobs to return (max: 100)
     * @param offset Offset for pagination
     * @param status Filter by status (completed, failed, processing, etc.)
     * @return Array of jobs
     * @throws ConvertorioException if request fails
     */
    public Job[] listJobs(int limit, int offset, String status) throws ConvertorioException {
        String endpoint = String.format("/v1/jobs?limit=%d&offset=%d", limit, offset);
        if (status != null && !status.isEmpty()) {
            endpoint += "&status=" + status;
        }

        JsonObject response = makeRequest("GET", endpoint, null);

        if (!response.get("success").getAsBoolean()) {
            throw new ConvertorioException(response.has("error") ?
                response.get("error").getAsString() : "Failed to list jobs");
        }

        return gson.fromJson(response.get("jobs"), Job[].class);
    }

    /**
     * Get job status
     *
     * @param jobId Job ID
     * @return Job details
     * @throws ConvertorioException if request fails
     */
    public Job getJob(String jobId) throws ConvertorioException {
        JsonObject response = makeRequest("GET", "/v1/jobs/" + jobId, null);

        if (!response.get("success").getAsBoolean()) {
            throw new ConvertorioException(response.has("error") ?
                response.get("error").getAsString() : "Failed to get job");
        }

        return gson.fromJson(response.get("job"), Job.class);
    }
}
